In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [6]:
data = pd.read_csv('Bhagvad Gita.csv')

In [7]:
data.head()  

Chapter No  Verse No                                             Shloka  \
0           1         1  धृतराष्ट्र उवाच | धर्मक्षेत्रे कुरुक्षेत्रे सम...   
1           1         2  सञ्जय उवाच । दृष्ट्वा तु पाण्डवानीकं व्यूढं दु...   
2           1         3  पश्यैतां पाण्डुपुत्राणामाचार्य महतीं चमूम् । व...   
3           1         4  अत्र शूरा महेष्वासा भीमार्जुनसमा युधि | युयुधा...   
4           1         5  धृष्टकेतुश्चेकितान: काशिराजश्च वीर्यवान् | पुर...   

                                 English Translation  \
0  Dhritarastra said: O Sanjaya, what did my sons...   
1  Sanjaya said: But then, seeing the army of the...   
2  O teacher, (please) see this vast army of the ...   
3  There are in this army, heroes wielding great ...   
4  Dhrstaketu, Cekitana, and the valiant king of ...   

                                         Explanation  
0  The two armies had gathered on the battlefield...  
1  Sanjay understood Dhritarashtra’s concern, who...  
2  Duryodhana asked Dronacharya to look at the sk...  
3  Due to his anxiety, the Pandava army seemed mu...  
4  Due to his anxiety, the Pandava army seemed mu...

In [9]:
data.describe(include='all') 

Chapter No    Verse No  \
count   700.000000  700.000000   
unique         NaN         NaN   
top            NaN         NaN   
freq           NaN         NaN   
mean      9.005714   23.224286   
std       5.589175   16.749599   
min       1.000000    1.000000   
25%       4.000000   10.000000   
50%       9.000000   20.000000   
75%      14.000000   32.000000   
max      18.000000   78.000000   

                                                   Shloka  \
count                                                 700   
unique                                                691   
top     अमी च त्वां धृतराष्ट्रस्य पुत्रा: सर्वे सहैवाव...   
freq                                                    2   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                      English Translation  \
count                                                 700   
unique                                                691   
top     And into You (enter) all those sons of Dhrtara...   
freq                                                    2   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                              Explanation  
count                                                 694  
unique                                                633  
top     To gain knowledge of the kṣhetra and kṣhetrajñ...  
freq                                                    5  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

The output shows the descriptive statistics of the DataFrame data, including the count, unique values, top value, and frequency for each column. The include='all' argument ensures that both categorical and numeric columns are included in the output.

For the numeric columns (Chapter No and Verse No), the output includes the count, mean, standard deviation, minimum value, 25th percentile, 50th percentile (median), 75th percentile, and maximum value.

For the categorical column (Shloka), the output includes the count, number of unique values, most frequent value (top), and its frequency.

For the other two columns (English Translation and Explanation), since they are not numeric or categorical, the output does not include any statistics for them. Instead, the output shows "NaN" (Not a Number) values for these columns.

Overall, this output provides a summary of the distribution of values in each column of the DataFrame, which can be helpful for data exploration and understanding.

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Chapter No           700 non-null    int64 
 1   Verse No             700 non-null    int64 
 2   Shloka               700 non-null    object
 3   English Translation  700 non-null    object
 4   Explanation          694 non-null    object
dtypes: int64(2), object(3)
memory usage: 27.5+ KB


This output shows information about the DataFrame, including the number of rows (700) and columns (5), as well as the column names, non-null count, and data type for each column.

The Chapter No and Verse No columns are integer data type with non-null count 700, which means all the rows have a value in these columns.

The Shloka, English Translation, and Explanation columns are object data type (i.e., string) with non-null count 700 for Shloka and English Translation, and 694 for Explanation. This means that there are 6 missing values in the Explanation column, as its non-null count is less than the total number of rows (700).

The memory usage line shows the amount of memory used by the DataFrame, which is approximately 27.5 KB. This can be useful to keep track of memory usage when working with large datasets.

In [12]:
#taking care of the missing values 
#every missing value i.e., is unique hence we have to maually replace them 
#we have only 6 missing value
#identifying the rows that contain missing values in the 'Explanation' olumn 
missing_values = data["Explanation"].isnull()


To know which rows have missing values in the Explanation column, you can use the isnull() method on the Explanation column and then use the sum() method to count the number of missing values in each column. This will print the total number of missing values in the Explanation column. 

In [14]:
missing_values = data["Explanation"].isnull().sum()
print(f"There are {missing_values} missing values in the Explanation column.")


There are 6 missing values in the Explanation column.


To see which rows have missing values, we can use boolean indexing to select those rows as following code. This will print all the rows that have missing values in the Explanation column, along with the values in the other columns.

In [15]:
missing_rows = data[data["Explanation"].isnull()]
print(missing_rows)


    Chapter No  Verse No                                             Shloka  \
7            1         8  भवान्भीष्मश्च कर्णश्च कृपश्च समितिञ्जय: | अश्व...   
8            1         9  अन्ये च बहव: शूरा मदर्थे त्यक्तजीविता: | नानाश...   
25           1        26  तत्रापश्यत्स्थितान् पार्थ: पितृ नथ पितामहान् |...   
41           1        42  सङ्करो नरकायैव कुलघ्नानां कुलस्य च | पतन्ति पि...   
42           1        43  दोषैरेतै: कुलघ्नानां वर्णसङ्करकारकै: | उत्साद्...   
43           1        44  उत्सन्नकुलधर्माणां मनुष्याणां जनार्दन | नरकेऽन...   

                                  English Translation Explanation  
7   (They are:) Your venerable self, Bhisma and Ka...         NaN  
8   There are many heroes who have dedicated their...         NaN  
25  Then Partha (Arjuna) saw, marshalled among bot...         NaN  
41  And the intermingling in the family leads the ...         NaN  
42  Due to these misdeeds of the ruiners of the fa...         NaN  
43  O Janardana, we have heard that li

This code assigns explanations for missing values in the "Explanation" column of the dataframe.

The first line defines a list of indices where there are missing values in the "Explanation" column.

The second line uses the loc function to select rows with those indices and the "Explanation" column, and assigns the corresponding list of explanations to them.

The first four explanations are assigned to the rows with indices 7, 8, 25, and 41, respectively. The fifth explanation is assigned to the row with index 42, and the last explanation is assigned to the row with index 43.

In [18]:
missing_values = [7, 8, 25, 41, 42, 43]

data.loc[missing_values, "Explanation"] = [    "There are personalities like yourself, Bheeshma, Karna, Kripa, Ashwatthama, Vikarn, and Bhurishrava, who are ever victorious in battle.",     "Also, there are many other heroic warriors, who are prepared to lay down their lives for my sake. They are all skilled in the art of warfare, and equipped with various kinds of weapons.",    "Missing explanation",    "An increase in unwanted children results in hellish life both for the family and for those who destroy the family. Deprived of sacrificial offerings, the ancestors of such corrupt families also fall.",    "Through the evil deeds of those who destroy the family tradition and thus give rise to unwanted progeny, a variety of social and family welfare activities are ruined.",    "O Janardan (Krishna), I have heard from the learned that those who destroy family traditions dwell in hell for an indefinite period of time."]


In [19]:
#checking for null or missing values 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Chapter No           700 non-null    int64 
 1   Verse No             700 non-null    int64 
 2   Shloka               700 non-null    object
 3   English Translation  700 non-null    object
 4   Explanation          700 non-null    object
dtypes: int64(2), object(3)
memory usage: 27.5+ KB


Here we can see there are no missing values in 'Explanation', i.e., from 694 -> 700

In [20]:
#checking it again in a different way 
# check for missing values
print(data.isnull())

# count the number of missing values in each column
print(data.isnull().sum())

     Chapter No  Verse No  Shloka  English Translation  Explanation
0         False     False   False                False        False
1         False     False   False                False        False
2         False     False   False                False        False
3         False     False   False                False        False
4         False     False   False                False        False
..          ...       ...     ...                  ...          ...
695       False     False   False                False        False
696       False     False   False                False        False
697       False     False   False                False        False
698       False     False   False                False        False
699       False     False   False                False        False

[700 rows x 5 columns]
Chapter No             0
Verse No               0
Shloka                 0
English Translation    0
Explanation            0
dtype: int64


In [21]:
#checking again for describing the dataset 
data.describe()

Chapter No    Verse No
count  700.000000  700.000000
mean     9.005714   23.224286
std      5.589175   16.749599
min      1.000000    1.000000
25%      4.000000   10.000000
50%      9.000000   20.000000
75%     14.000000   32.000000
max     18.000000   78.000000

This will save the edited DataFrame to a new csv file called "edited_data.csv" in the same directory as your Python script or notebook. The index=False argument specifies that the index column should not be included in the saved csv.

In [22]:
#saving the edited data
data.to_csv('edited_data.csv', index=False)